net/instaweb/rewriter/csp.cc

```diff
@@ -49,43 +49,139 @@ inline bool IsAsciiAlpha(char ch) {
           ((ch >= 'A') && (ch <= 'Z')));
 }
 
+inline bool IsAsciiDigit(char ch) {
+  return ((ch >= '0') && (ch <= '9'));
+}
+
 }  // namespace
 
+bool CspSourceExpression::TryParseScheme(StringPiece* input) {
+  if (input->size() < 2) {
+    // Need at least a: or such.
+    return false;
+  }
+
+  if (!IsAsciiAlpha((*input)[0])) {
+    return false;
+  }
+
+  size_t pos = 1;
+  while (pos < input->size() &&
+         (IsAsciiAlphaNumeric((*input)[pos])
+          || (*input)[pos] == '+'
+          || (*input)[pos] == '-'
+          || (*input)[pos] == '.')) {
+    ++pos;
+  }
+
+  if (pos == input->size() || (*input)[pos] != ':') {
+    // All schema characters, but no : or :// afterwards -> something else.
+    return false;
+  }
+
+  if (pos == (input->size() - 1)) {
+    // Last character was also : -> clearly a scheme-source
+    kind_ = kSchemeSource;
+    input->substr(0, pos).CopyToString(&mutable_url_data()->scheme_part);
+    input->remove_prefix(pos + 1);
+    return true;
+  }
+
+  // We now want to see if it's actually foo://
+  if ((pos + 2 < input->size())
+       && ((*input)[pos + 1] == '/') && ((*input)[pos + 2] == '/')) {
+    input->substr(0, pos).CopyToString(&mutable_url_data()->scheme_part);
+    input->remove_prefix(pos + 3);
+  }
+
+  // Either way, it's not a valid scheme-source at this point, even if it's
+  // a valid host-source
+  return false;
+}
+
 CspSourceExpression CspSourceExpression::Parse(StringPiece input) {
   TrimCspWhitespace(&input);
   if (input.empty()) {
-    return CspSourceExpression(kUnknown);
+    return CspSourceExpression();
```

Comment:
I wonder, will we need to be able to discern between unknown and empty/whitespace ?